File: src/drone_description/src/teleop_drone.cpp
#include <ros/ros.h>
#include <std_msgs/Float64MultiArray.h>
#include <termios.h>
#include <unistd.h>
#include <sys/select.h>
#include <map>
#include <array>
#include <vector>

// key to delta speeds
static std::map<char, std::array<double,4>> bindings = {
    {'w', { 10,  10,  10,  10}},
    {'s', {-10, -10, -10, -10}},
    {'i', { 10, -10, -10,  10}},
    {'k', {-10,  10,  10, -10}},
    {'j', {-10, -10,  10,  10}},
    {'l', { 10,  10, -10, -10}},
    {'a', { 10, -10,  10, -10}},
    {'d', {-10,  10, -10,  10}}
};

// Read one character without waiting for Enter
int getch(){
    static struct termios oldt, newt;
    tcgetattr(STDIN_FILENO, &oldt);
    newt = oldt;
    newt.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, &newt);
    int c = getchar();
    tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
    return c;
}

int main(int argc, char** argv){
    ros::init(argc, argv, "drone_teleop_cpp");
    ros::NodeHandle nh;
    ros::Publisher pub = nh.advertise<std_msgs::Float64MultiArray>("/motor_speed", 1);

    std::vector<double> speeds(4, 0.0);
    ROS_INFO("\nControls:\n  w/s : collective throttle\n  i/k : pitch forward/back\n  j/l : roll left/right\n  a/d : yaw left/right\nCTRL-C to quit\n");

    ros::Rate rate(20);
    while(ros::ok()){
        // wait up to 10 ms for a key
        struct timeval timeout;
        timeout.tv_sec  = 0;
        timeout.tv_usec = 10000;

        if(::isatty(STDIN_FILENO) &&
           ::select(STDIN_FILENO+1, nullptr, nullptr, nullptr, &timeout) > 0)
        {
            char c = getch();
            if(c=='\x03') break;  // Ctrl-C
            auto it = bindings.find(c);
            if(it != bindings.end()){
                auto &d = it->second;
                for(int i=0; i<4; ++i){
                    speeds[i] = std::max(0.0, speeds[i] + d[i]);
                }
            }
        }

        std_msgs::Float64MultiArray msg;
        msg.data = speeds;
        pub.publish(msg);

        ros::spinOnce();
        rate.sleep();
    }

    // stop on exit
    std_msgs::Float64MultiArray stop_msg;
    stop_msg.data = {0,0,0,0};
    pub.publish(stop_msg);

    return 0;
}

File: src/drone_description/scripts/teleop_drone.py
#!/usr/bin/env python3
import rospy, sys, select, termios, tty
from std_msgs.msg import Float64MultiArray

# Key → (Δ front_right, Δ front_left, Δ rear_left, Δ rear_right)
bindings = {
    'w': ( 10,  10,  10,  10),   # all props up
    's': (-10, -10, -10, -10),   # all props down
    'i': ( 10, -10, -10,  10),   # pitch forward
    'k': (-10,  10,  10, -10),   # pitch back
    'j': (-10, -10,  10,  10),   # roll left
    'l': ( 10,  10, -10, -10),   # roll right
    'a': ( 10, -10,  10, -10),   # yaw left
    'd': (-10,  10, -10,  10),   # yaw right
}

msg = """
Controls:
  w/s : collective throttle
  i/k : pitch forward/back
  j/l : roll left/right
  a/d : yaw left/right
CTRL-C to quit
"""

def get_key():
    fd = sys.stdin.fileno()
    old = termios.tcgetattr(fd)
    try:
        tty.setraw(fd)
        if select.select([sys.stdin], [], [], 0.1)[0]:
            return sys.stdin.read(1)
        return ''
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old)

def teleop():
    rospy.init_node('drone_teleop')
    pub = rospy.Publisher('/motor_speed', Float64MultiArray, queue_size=1)
    settings = termios.tcgetattr(sys.stdin)
    speeds = [0.0, 0.0, 0.0, 0.0]

    try:
        print(msg)
        rate = rospy.Rate(20)
        while not rospy.is_shutdown():
            key = get_key()
            if key == '\x03':  # Ctrl+C
                break

            if key in bindings:
                delta = bindings[key]
                # increment each rotor’s speed
                speeds = [max(0.0, sp + d) for sp, d in zip(speeds, delta)]

            # publish array of four speeds
            pub.publish(Float64MultiArray(data=speeds))
            rate.sleep()

    except rospy.ROSInterruptException:
        pass
    finally:
        # zero them out on exit
        pub.publish(Float64MultiArray(data=[0,0,0,0]))
        termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)

if __name__ == '__main__':
    teleop()
